### PR TITLE
fix(disperser-client): RetrieveBlob grpc max size regression bug

### PR DESCRIPTION
**Pull request by samlaf**
A previous commit 6894828 (https://github.com/Layr-Labs/eigenda/pull/826) caused regression by forgetting to set grpc max msg size in retrieveBlob call

Caught by integration test from eigenda-proxy: https://github.com/Layr-Labs/eigenda-proxy/actions/runs/11584563970/job/32251942842?pr=196

Added unit-test here to catch any future regression on this feature. Also made that connection size a parameter of the disperser-client (keeps the default 100MiB that was previously set).

## Checks

- [ ] I've made sure the lint is passing in this PR.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, in that case, please comment that they are not relevant.
- [ ] I've checked the new test coverage and the coverage percentage didn't drop.
- Testing Strategy
   - [x] Unit tests
   - [ ] Integration tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: api/clients/disperser_client.go [hunk 1]
```diff
@@ -26,6 +26,14 @@ type Config struct {
 	// TODO: do we want to add config timeouts for those separate requests?
 	Timeout           time.Duration
 	UseSecureGrpcFlag bool
+	// MaxRetrieveBlobSizeBytes is the maximum size of a blob that can be retrieved by using
+	// the RetrieveBlob method. This is used to set the max message size for the grpc client.
+	// DisperserClient uses a single underlying grpc channel shared for all methods,
+	// but all other methods use the default 4MiB max message size, whereas RetrieveBlob
```
Comment (bxue-l2) [inline api/clients/disperser_client.go:32]: I think dispersal and retrieval for a blob have the same blob length.

It is just the receiving ends are different. On our dispersal endpoint, we also set a very large buffer size. Then similarly, if you retrieve, you should set a large buffer. The buffer size on the sender side does not matter

File: api/clients/disperser_client.go [hunk 2]
```diff
@@ -38,7 +46,7 @@ func NewConfig(hostname, port string, timeout time.Duration, useSecureGrpcFlag b
 	}
 }
 
-type DisperserClient interface {
+type IDisperserClient interface {
```
Comment (bxue-l2) [inline api/clients/disperser_client.go:49]: I don't think we use this convention, this looks solidity

 what do you think @ian-shim

File: api/clients/disperser_client.go [hunk 3]
```diff
@@ -273,18 +284,25 @@ func (c *disperserClient) GetBlobStatus(ctx context.Context, requestID []byte) (
...
 		return nil, api.NewErrorInternal(err.Error())
 	}
 
 	ctxTimeout, cancel := context.WithTimeout(ctx, time.Second*60)
 	defer cancel()
-	reply, err := c.client.RetrieveBlob(ctxTimeout, &disperser_rpc.RetrieveBlobRequest{
-		BatchHeaderHash: batchHeaderHash,
-		BlobIndex:       blobIndex,
-	})
+	if c.config.MaxRetrieveBlobSizeBytes == 0 {
```
Comment (bxue-l2) [inline api/clients/disperser_client.go:295]: should we have a flag for this? this looks adhoc

File: api/clients/disperser_client.go [hunk 4]
```diff
@@ -273,18 +284,25 @@ func (c *disperserClient) GetBlobStatus(ctx context.Context, requestID []byte) (
...
+		// max blob size on mainnet is currently 16MiB but we set this to 100MiB for backward compatibility
```
Comment (pschork) [inline api/clients/disperser_client.go:296]: @samlaf are you planning to update this comment every time the disperser's `MAX_BLOB_SIZE` config is changed? I would rephrase this to `set to 100MiB for forward compatibility. check official documentation for current max blob size on mainnet`

File: api/clients/disperser_client.go [hunk 5]
```diff
@@ -26,6 +26,14 @@ type Config struct {
 	// TODO: do we want to add config timeouts for those separate requests?
 	Timeout           time.Duration
 	UseSecureGrpcFlag bool
+	// MaxRetrieveBlobSizeBytes is the maximum size of a blob that can be retrieved by using
+	// the RetrieveBlob method. This is used to set the max message size for the grpc client.
+	// DisperserClient uses a single underlying grpc channel shared for all methods,
+	// but all other methods use the default 4MiB max message size, whereas RetrieveBlob
+	// potentially needs a larger size.
+	//
+	// If not set, the default value is 100MiB, to be backward compatible.
```
Comment (pschork) [inline api/clients/disperser_client.go:35]: forward compatible

Comment (samlaf) [reply api/clients/disperser_client.go:35]: It's actually backward compatible. The code before my previous PR (that caused the regression) had a default 100MiB.

File: api/clients/disperser_client.go (hunk 2 above)
Comment (samlaf) [reply api/clients/disperser_client.go:49]: I know, I don't like it either, but we already use this for eigenda_client, so I just followed the convention there.
Was forced to make the `disperserClient` struct public for the tests, so had to find another name for the interface. More golangy might be `DisperserClienter`. Honestly I think defining the interface here is code smell and should eventually all be refactored. Interfaces should be defined case-by-case at the call sites.

File: api/clients/disperser_client.go (hunk 1 above)
Comment (samlaf) [reply api/clients/disperser_client.go:32]: It does, both sides set their respective length that they want.
image

Disperser needs large buffer to receive large blobs for dispersal, and client needs large buffer size to retrieve blobs.

File: api/clients/disperser_client.go (hunk 5 above)
Comment (samlaf) [reply api/clients/disperser_client.go:35]: Ah just saw your previous comment, I see what you mean now! Good point

Comment (samlaf) [reply api/clients/disperser_client.go]: [a07336c](https://github.com/Layr-Labs/eigenda/pull/849/commits/a07336cf2fce85b7d1cf886a3d07f3306ac459b8)

File: api/clients/disperser_client.go (hunk 4 above)
Comment (samlaf) [reply api/clients/disperser_client.go]: Good call: [a07336c](https://github.com/Layr-Labs/eigenda/pull/849/commits/a07336cf2fce85b7d1cf886a3d07f3306ac459b8)

File: api/clients/disperser_client.go (hunk 3 above)
Comment (bxue-l2) [reply api/clients/disperser_client.go:296]: commented in the wrong place, make c.config.MaxRetrieveBlobSizeBytes set in the constructor with the flag, so no if statement here

File: api/clients/disperser_client.go (hunk 2 above)
Comment (bxue-l2) [reply api/clients/disperser_client.go:50]: Ok, saw it. we used IEigenDAClient.

Looks fine to me

File: api/clients/disperser_client.go (hunk 3 above)
Comment (bxue-l2) [reply api/clients/disperser_client.go:296]: I mean MaxRetrieveBlobSizeBytes is always 0. since there is no flag to set it

File: api/clients/disperser_client.go (hunk 2 above)
Comment (ian-shim) [reply api/clients/disperser_client.go:50]: > Was forced to make the disperserClient struct public for the tests

Don't see any place that uses it. We couldn't use the constructor in the test?

File: api/clients/disperser_client.go (hunk 3 above)
Comment (samlaf) [reply api/clients/disperser_client.go:296]: It's not. constructor takes a config, which could have set MaxRetrieveBlobSizeBytes
```
func NewDisperserClient(config *Config, signer core.BlobRequestSigner) *DisperserClient {
```

And as I said above, I prefer not to set in constructor, because one can still build a DisperserClient without using the constructor. This is often done in tests for eg, like
```
client := &DisperserClient{
   Conn: ...
   Foo: ...
   
```

Comment (samlaf) [reply api/clients/disperser_client.go:296]: Ahhhh scratch that... fields are private, so constructor actually has to be used. Ok, you're right, will update

Comment (samlaf) [reply api/clients/disperser_client.go]: [04499a5](https://github.com/Layr-Labs/eigenda/pull/849/commits/04499a54c72071151d1bee67b1ba8e68d98328c6)

**Review by bxue-l2 — APPROVED** [review]

**Comment by bxue-l2** [comment]
lgtm, but address Ian's comment

File: api/clients/disperser_client.go (hunk 2 above)
Comment (samlaf) [reply api/clients/disperser_client.go:50]: Ah, you're right. Originally was trying to put the tests in the same package so I can build the struct manually injecting the connection to test something. But then realized I can just add the config parameter.

Comment (samlaf) [reply api/clients/disperser_client.go:50]: So hmmm... could revert to the old names. But somehow I feel like our clients should follow the same convention.

Comment (samlaf) [reply api/clients/disperser_client.go]: Reverted: [dc98581](https://github.com/Layr-Labs/eigenda/pull/849/commits/dc985811c0bdc2d47e4e046c67b4756294258505)
Also added TODO to change eigenda-client to use same naming convention as disperser-client: [57d716f](https://github.com/Layr-Labs/eigenda/pull/849/commits/57d716fec20db630599a969ce1b4b7ecbe61242c)

File: api/clients/disperser_client.go [hunk 6]
```diff
@@ -91,6 +103,14 @@ var _ DisperserClient = &disperserClient{}
 //	// Subsequent calls will use the existing connection
 //	status2, requestId2, err := client.DisperseBlob(ctx, otherData, otherQuorums)
 func NewDisperserClient(config *Config, signer core.BlobRequestSigner) *disperserClient {
+	if config == nil {
+		config = &Config{}
```
Comment (ian-shim) [inline api/clients/disperser_client.go:107]: should it return error if config is nil?

**Review by ian-shim — APPROVED** [review]
lgtm, left one more comment!

File: api/clients/disperser_client.go (hunk 6 above)
Comment (samlaf) [reply api/clients/disperser_client.go]: yep good call: [f2100c8](https://github.com/Layr-Labs/eigenda/pull/849/commits/f2100c8f0055c27ac0678ae54e8f8c3eccc377cd)
